### ISSUE TITLE
Performance problem with high concurrent S3 requests

### ISSUE THREAD

**Issue by MannerMan** [chrislusf/seaweedfs#2145]
**Some background:** 
We have a 36-node (volume and filers with s3) seaweed deployment, with 1 separate master server. We use HAproxy in front of the cluster, and loadbalance S3 requests to all 36 filers. We use postgres as filer storage (with the regular 'postgres' connector, not 'postgres2'), on a separate dedicated server with 16 cores and 64 gb of ram. Our filer meta storage currently uses up about 150gb on disk. We store our postgres backups for SaaS product in seaweed via S3 by using [pgbackrest](https://github.com/pgbackrest/pgbackrest) backup software. Our postgres database servers for the SaaS application have an uncommon data-model where we end up with about 5 million files in our postgres data directory for each server. Since pgbackrest does file-level backup, we choose seaweed as a fitting storage for the many many millions of files pgbackrest produces.

We appear to have encountered a performance problem however, and are trying to figure out where the bottleneck might be. We can successfully backup one server in about 3 hours if no other backups are running. But when we attempt to start many simultaneous backups with pgbackrest (about 15-20 concurrently), we run into some kind of congestion issue somewhere in the system. 

**Describe the bug**
While running many concurrent S3 requests, some kind of scalability issue occurs. CPU, memory and I/O does not seem to be an issue on any node as far as we can tell (including filer meta storage server and master server). 
Might be easier to visualize the problem in graphs;
![Screenshot from 2021-06-21 14-00-39](https://user-images.githubusercontent.com/1773068/122760166-1b04df00-d29b-11eb-9f8d-170edda6e91c.png)
![Screenshot from 2021-06-21 14-00-46](https://user-images.githubusercontent.com/1773068/122760172-1ccea280-d29b-11eb-8cf3-c90962693604.png)
![Screenshot from 2021-06-21 14-00-53](https://user-images.githubusercontent.com/1773068/122760180-1e986600-d29b-11eb-8fc5-d797fe07707e.png)
![Screenshot from 2021-06-21 14-01-00](https://user-images.githubusercontent.com/1773068/122760184-1fc99300-d29b-11eb-9f60-5ff61444eab2.png)

At around 20:30 one can see backups begins to start up on postgres servers, we have some random sleep implemented so they wont all start at the same time. As seen in the graphs, we have a steady S3 QPS increase up to about 100 per node until about 22:30 when it just tanks down to 40 and stays there for most of the night while backups are running. At 05:15, backups begin to fail (job has 6h timeout) and we then see a short interval where QPS increase again. At 06:00 all backups have failed due to timeout.
We think these graphs suggest there is some congestion issue; many concurrent requests/connections causing some bottleneck, since at 05:15 when backups begin to fail, we get a QPS increase.

**System Setup**
- List the command line to start "weed master", "weed volume", "weed filer", "weed s3", "weed mount".
 * master: `/usr/local/sbin/weed master -ip.bind 0.0.0.0 -ip 10.84.101.100 -mdir /opt/seaweedfs/seaweed-backup001.prod.int/master -defaultReplication 001`
 * volume: `/usr/local/sbin/weed volume -port 8080 -metricsPort 9327 -publicUrl http://seaweed-backup001.prod.int -max 32 -ip.bind 0.0.0.0 -ip 10.84.101.101 -dir /opt/seaweedfs/seaweed-backup001.prod.int/volume -mserver 10.84.101.100:9333 -dataCenter sto4 -rack rack1`
 * filer/s3 `/usr/local/sbin/weed filer -port 8889 -metricsPort 9328 -ip.bind 0.0.0.0 -ip 10.84.101.101 -s3 -master 10.84.101.100:9333 -dataCenter=sto4 -rack=rack1 -s3.config=s3_config.json`
- OS version: CentOS 8.4
- output of `weed version`: `version 30GB 2.52 76f24af linux amd64`
- if using filer, show the content of `filer.toml`
```
[postgres]
enabled = true

hostname = "pg-backup001-db001.prod.int"
port = 6432
username = "seaweedfiler"
password = "<redacted>"
database = "seaweedfiler"          # create or use an existing database
sslmode = "require"
connection_max_idle = 2
connection_max_open = 10

# if insert/upsert failing, you can disable upsert or update query syntax to match your RDBMS syntax:
enableUpsert = true
upsertQuery = """INSERT INTO "%[1]s" (dirhash,name,directory,meta) VALUES($1,$2,$3,$4) ON CONFLICT (dirhash,name) DO UPDATE SET meta = EXCLUDED.meta WHERE "%[1]s".meta != EXCLUDED.meta""
```

**Questions**
* Any tips how we can track down the bottleneck or where to look?
* Which filer meta storage is best for high-concurrency and scalability? Cassandra? Redis and Etcd does not seem fitting due to the amount of data. We're suspecting that postgres might be the component causing the bottleneck. We reach about 20k TPS at peak, but it also drops with the weird thing happening at 22:30.
* Is it better to run only a few filers (3-5?) with s3 rather than having s3 enabled on all 36 nodes?
* If changing filer meta storage, does starting a new filer with a different filer store (for example cassandra) using `-peers=` option synchronize existing metadata from old filer storage to the new filer, or does it only insert new incoming metadata?

**Comment by kmlebedev** [comment]
@MannerMan  Try set option  -concurrentUploadLimitMB equal to the available memory

**Comment by PeterCxy** [comment]
I would like to add that running the same requests over filter itself or through FUSE mount does not seem to trigger the same performance issue at the same load level as with the S3 gateway.

**Comment by chrislusf** [comment]
Maybe check how is the number of open files?

**Comment by jonaz** [comment]
Hi, i work with @MannerMan and we dont se any increase in openfiles on any seaweed nodes. Our hardlimit is 262144 and we are not close at hitting that. 
![2021-06-22-081453_1394x665_scrot](https://user-images.githubusercontent.com/1146389/122873547-31aa4500-d332-11eb-8847-b0c2e97e9fb9.png)

**Comment by jonaz** [comment]
It was not an openfile problem. Same result after increase as i suspected. We'll try upping -concurrentUploadLimitMB before next backup scheduled run.

**Comment by MannerMan** [comment]
This issue can probably be closed now. We troubleshoot the problem using the benchmark tool hsbench mentioned in https://github.com/chrislusf/seaweedfs/wiki/S3-API-Benchmark and loaded seaweed s3 with hundreds of threads to simulate our backup software. This made it much easier to troubleshoot since we could quickly iterate on fixes. Performance turned out to be bottlenecked by a combination of things, mainly ;
* The seaweed master node was cpu-overloaded (it had 4 cpu cores, now increased to 16). 
* We used pgbouncer in-front of postgres for filer meta store, but that appears to have lowered performance in this case due to it being single threaded, we now connected filer directly to postgres instead. 
* We set connection_max_idle and connection_max_open to 10/10, previously it was at 2/10 which caused seaweed to disconnect/connect to postgres very frequently lowering performance.
* We beefed up our postgres meta storage with more cores and tuned configuration to handle more connections and concurrent writes. 
* We set `-idleTimeout=300` in volume server and `-concurrentUploadLimitMB=2048 -maxMB=32` in filer
* Our loadbalancer was missing 12 of the filer nodes, thereby only loadbalcing to 2/3 of the cluster

All in all, we increased S3 requests per second from about 2k up to 8k in the benchmark with all fixes combined, this was enough to ensure our backups now finish in time!
